component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/services/TOTPAdminService.java

```diff
@@ -61,6 +61,89 @@ public String initTOTP(String username, AuthenticationContext context) throws TO
 		return TOTPKeyGenerator.addTOTPClaimsAndRetrievingQRCodeURL(claims, username, context);
 	}
 
+	/**
+	 * Generate TOTP secret for a given user and will be saved in http://wso2.org/claims/identity/verifySecretkey claim.
+	 *
+	 * @param username Username of the user
+	 * @return Encoded QR Code URL.
+	 * @throws TOTPException
+	 */
+	public String generateSecret(String username) throws TOTPException {
+
+		Map<String, String> claims = TOTPKeyGenerator.generateClaims(username, false);
+
+		String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+		try {
+			UserRealm userRealm = TOTPUtil.getUserRealm(username);
+			if (userRealm != null) {
+				Map<String, String> claimsToPersist = new HashMap<>();
+				claimsToPersist.put(TOTPAuthenticatorConstants.VERIFY_SECRET_KEY_CLAIM_URL,
+						claims.get(TOTPAuthenticatorConstants.SECRET_KEY_CLAIM_URL));
+				userRealm.getUserStoreManager().setUserClaimValues(tenantAwareUsername, claimsToPersist, null);
+			}
+		} catch (UserStoreException e) {
+			throw new TOTPException(
+					"Failed to access user store manager to store secret key for the user: " + tenantAwareUsername, e);
+		} catch (AuthenticationFailedException e) {
+			throw new TOTPException("Error while retrieving the user realm for the user: " + tenantAwareUsername, e);
+		}
+
+		return claims.get(TOTPAuthenticatorConstants.QR_CODE_CLAIM_URL);
+	}
+
+	/**
+	 * Enable the TOTP for the given user. OTP will be validated against the secret saved in
+	 * http://wso2.org/claims/identity/verifySecretkey claim and if successful, secret is moved to
+	 * http://wso2.org/claims/identity/secretkey claim.
+	 *
+	 * @param username            Username of the user
+	 * @param verificationCode    verification code generated from the secret issued with {@link #generateSecret(String)}
+	 * @return if TOTP enabling is successful or not.
+	 * @throws TOTPException
+	 */
+	public boolean enableTOTP(String username, int verificationCode) throws TOTPException {
+
+		String tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(username);
+		try {
+			UserRealm userRealm = TOTPUtil.getUserRealm(username);
+			if (userRealm != null) {
```

Comment:
Can there be instances where the user realm in null?
If so shall we add a debug log in the else block.